src/redux/utils.js

```diff
@@ -97,93 +78,111 @@ export const getNewAccountData = async (dispatch, icpPrice, state) => {
+  const { details } = transaction;
+  const { fee } = details;
+
+  const parsedDetails = parseTransactionObject(details);
+  let parsedFee = fee;
+
+  if (fee instanceof Object && ('token' in fee || 'currency' in fee)) {
+    parsedFee = parseTransactionObject(fee);
+  }
+
+  return {
+    ...transaction,
+    details: {
+      ...parsedDetails,
+      fee: parsedFee,
+    },
+  };
+};
+
+const getTransactionSymbol = details => {
+  if (!details) {
+    return '';
+  }
+  if ('tokenRegistryInfo' in (details?.canisterInfo || [])) {
+    return details?.canisterInfo.tokenRegistryInfo.symbol;
+  }
+  if ('nftRegistryInfo' in (details?.canisterInfo || [])) {
+    return 'NFT';
+  }
+  return (
+    details?.currency?.symbol ??
+    details?.sonicData?.token?.details?.symbol ??
+    details?.details?.name ??
+    ''
+  );
+};
+
+const getTransactionType = (type, isOwnTx) => {
+  if (!type) {
+    return '';
+  }
+  if (type.includes('transfer')) {
+    return isOwnTx ? 'SEND' : 'RECEIVE';
+  }
+  if (type.includes('Liquidity')) {
+    return `${type.includes('removeLiquidity') ? 'Remove' : 'Add'} Liquidity`;
+  }
+  return type.toUpperCase();
+};
+
+export const mapTransaction = (icpPrice, state) => trx => {
+  const { principal, accountId } = state.keyring?.currentWallet;
+
+  let parsedTransaction = recursiveParseBigint(parseTransaction(trx));
+  const { details, hash, caller, timestamp } = parsedTransaction || {};
+  const isOwnTx = [principal, accountId].includes(caller);
+
+  const symbol = getTransactionSymbol(details);
+  const asset = formatAssetBySymbol(details?.amount, symbol, icpPrice);
+  const type = getTransactionType(parsedTransaction?.type, isOwnTx);
 
   const transaction = {
-    ...asset,
-    type: getType(),
-    hash: trx?.hash,
-    to: trx?.details?.to,
-    from: trx?.details?.from || trx?.caller,
-    date: new Date(trx?.timestamp),
-    status: ACTIVITY_STATUS[trx?.details?.status],
-    image:
-      trx?.details?.details?.icon ||
-      trx?.details?.icon ||
-      TOKEN_IMAGES[symbol] ||
-      canisterInfo?.icon ||
-      '',
+    amount: asset.amount,
+    value: asset.value,
+    icon: asset.icon,
+    type,
+    hash,
+    to: details?.to?.principal,
+    from: details?.from?.principal || caller,
+    date: new Date(timestamp),
+    status: ACTIVITY_STATUS[details?.status],
+    image: details?.canisterInfo?.icon || TOKEN_IMAGES[symbol] || '',
     symbol,
-    canisterId: trx?.details?.canisterId,
+    canisterId: details?.canisterId,
     plug: null,
-    canisterInfo,
-    details: { ...trx?.details, caller: trx?.caller },
+    canisterInfo: details?.canisterInfo,
+    details: {
+      ...details,
+      caller,
+    },
   };
   return transaction;
 };
 
 export const formatContact = contact => ({
   image: contact.emoji[0],
   name: contact.name,
-  id: contact.value?.PrincipalId,
+  id: contact.value?.PrincipalId?.toText(), //TODO Check this logic. What happens if principal doesnt come from the contact?
```

Comment:
hm that's a good q. we should ask juan. AFAIK it could be one of 3 things: pid, accid or icns